torch/distributed/checkpoint/state_dict.py

```diff
@@ -332,6 +344,130 @@ def _init_optim_state(optim: torch.optim.Optimizer) -> None:
     optim.zero_grad(set_to_none=True)
 
 
+def _flatten_optim_state_dict(value: Any, prefix: str = "") -> Dict[str, ValueType]:
+    ret = {}
+    if isinstance(value, dict):
+        for key, child_value in value.items():
+            next_prefix = f"{prefix}.{key}" if prefix else key
+            ret.update(_flatten_optim_state_dict(child_value, next_prefix))
+    elif isinstance(value, (DTensor, ShardedTensor, torch.Tensor, int, float)):
+        ret[prefix] = value
+    elif isinstance(value, (list, tuple)):
+        assert len(value) > 0, prefix
+        for v in value:
+            assert isinstance(v, dict)
```

Comment:
I think that this code would work on the state dicts of our current optimizers, but is not general enough to handle user-added state. For example, we allow users to add their own state to state in the optimizer so they could track whatever they want. I can therefore conceive of people adding an entry to state that is a list of scalars vs a list of dictionaries as expected here.

One use case that could be conceivable is users tacking on a list of learning rates on different devices, let's say, and then using the right lr per device in the optimizer.step() which is something @mlazos had brought up to me. It is true users could perhaps reuse param_groups for something like this, but my general point is that we currently do not limit how users can tack things onto optimizer state, so we can't assert that every optimizer state will only have lists/tuples of dictionaries.

Would it be difficult to handle more general cases? (Either by flattening them according to index, etc, or by leaving them as is?) Either way, we should be blatant about what this supports vs not.